src/binder_data.c

```diff
@@ -0,0 +1,2452 @@
+    call->addresses = g_strsplit(dc->addresses.data.str, " ", -1);
+    call->pcscf = g_strsplit(dc->pcscf.data.str, " ", -1);
+
+    DBG("[status=%d,retry=%d,cid=%d,active=%d,type=%s,ifname=%s,"
+        "mtu=%d,address=%s,dns=%s,gateways=%s,pcscf=%s]",
+        call->status, call->retry_time, call->cid, call->active,
+        dc->type.data.str, call->ifname, call->mtu, dc->addresses.data.str,
+        dc->dnses.data.str, dc->gateways.data.str, dc->pcscf.data.str);
+    return call;
+}
+
+static
+GSList*
+binder_data_call_list_1_0(
+    const RadioDataCall* calls,
+    gsize n)
+{
+    if (n) {
+        gsize i;
+        GSList* l = NULL;
+
+        DBG("num=%u", (guint) n);
+        for (i = 0; i < n; i++) {
+            l = g_slist_insert_sorted(l, binder_data_call_new_1_0(calls + i),
+                binder_data_call_compare);
+        }
+        return l;
+    } else {
+        DBG("no data calls");
+        return NULL;
+    }
+}
+
+static
+BinderDataCall*
+binder_data_call_new_1_4(
+    const RadioDataCall_1_4* dc)
+{
+    BinderDataCall* call = binder_data_call_new();
+
+    call->cid = dc->cid;
+    call->status = dc->cause;
+    call->active = dc->active;
+    call->prot = dc->type;
+    call->retry_time = dc->suggestedRetryTime;
+    call->mtu = dc->mtu;
+    call->ifname = g_strdup(dc->ifname.data.str);
+    call->dnses = binder_strv_from_hidl_string_vec(&dc->dnses);
+    call->gateways = binder_strv_from_hidl_string_vec(&dc->gateways);
+    call->addresses = binder_strv_from_hidl_string_vec(&dc->addresses);
+    call->pcscf = binder_strv_from_hidl_string_vec(&dc->pcscf);
+
+    DBG("[status=%d,retry=%d,cid=%d,active=%d,type=%d,ifname=%s,"
+        "mtu=%d,address=%s,dns=%s,gateways=%s,pcscf=%s]",
+        call->status, call->retry_time, call->cid, call->active,
+        dc->type, call->ifname, call->mtu,
+        binder_print_strv(call->addresses, " "),
+        binder_print_strv(call->dnses, " "),
+        binder_print_strv(call->gateways, " "),
+        binder_print_strv(call->pcscf, " "));
+    return call;
+}
+
+static
+GSList*
+binder_data_call_list_1_4(
+    const RadioDataCall_1_4* calls,
+    gsize n)
+{
+    if (n) {
+        gsize i;
+        GSList* l = NULL;
+
+        DBG("num=%u", (guint) n);
+        for (i = 0; i < n; i++) {
+            l = g_slist_insert_sorted(l, binder_data_call_new_1_4(calls + i),
+                binder_data_call_compare);
+        }
+        return l;
+    } else {
+        DBG("no data calls");
+        return NULL;
+    }
+}
+
+static
+gboolean
+binder_data_call_equal(
+    const BinderDataCall* c1,
+    const BinderDataCall* c2)
+{
+    if (!c1 && !c2) {
+        return TRUE;
+    } else if (c1 && c2) {
+        return c1->cid == c2->cid &&
+            c1->status == c2->status &&
+            c1->active == c2->active &&
+            c1->prot == c2->prot &&
+            c1->retry_time == c2->retry_time &&
+            c1->mtu == c2->mtu &&
+            !g_strcmp0(c1->ifname, c2->ifname) &&
```

Comment:
It's `g_strdup`'ped string from the incoming transaction, I've never seen a NULL string coming out of binder. Even `strcmp` would work here I think, `g_strcmp0` is there just in case))